clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractPartitionAssignor.java

```diff
@@ -0,0 +1,183 @@
+ * file distributed with this work for additional information regarding copyright ownership. The ASF licenses this file
+ * to You under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the
+ * License. You may obtain a copy of the License at
+ * <p/>
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * <p/>
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+package org.apache.kafka.clients.consumer.internals;
+
+import org.apache.kafka.clients.consumer.PartitionAssignor;
+import org.apache.kafka.common.TopicPartition;
+import org.apache.kafka.common.protocol.types.ArrayOf;
+import org.apache.kafka.common.protocol.types.Field;
+import org.apache.kafka.common.protocol.types.Schema;
+import org.apache.kafka.common.protocol.types.SchemaException;
+import org.apache.kafka.common.protocol.types.Struct;
+import org.apache.kafka.common.protocol.types.Type;
+import org.apache.kafka.common.utils.Utils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+public abstract class AbstractPartitionAssignor implements PartitionAssignor<AbstractPartitionAssignor.ConsumerMetadata> {
+    private static final String SUBSCRIPTION_KEY_NAME = "subscription";
+    private static final String TOPIC_KEY_NAME = "topics";
+    private static final String TOPIC_PATTERN_KEY_NAME = "topic_pattern";
+    private static final String METADATA_HASH_KEY_NAME = "metadata_hash";
+
+    public static final Schema CONSUMER_METADATA_SUBSCRIPTION_V0 = new Schema(
+            new Field(TOPIC_KEY_NAME, new ArrayOf(Type.STRING)),
+            new Field(TOPIC_PATTERN_KEY_NAME, Type.STRING),
+            new Field(METADATA_HASH_KEY_NAME, Type.BYTES));
+
+    public static final Schema CONSUMER_METADATA_V0 = new Schema(
+            new Field(SUBSCRIPTION_KEY_NAME, CONSUMER_METADATA_SUBSCRIPTION_V0));
+
+    private final Logger log = LoggerFactory.getLogger(getClass());
+
+    public abstract List<TopicPartition> assign(String consumerId,
+                                                Map<String, Integer> partitionsPerTopic,
+                                                Map<String, ConsumerMetadata> consumerMetadata);
+
+    public Type schema() {
+        return new Type() {
+            @Override
+            public void write(ByteBuffer buffer, Object o) {
+                ConsumerMetadata metadata = (ConsumerMetadata) o;
+                CONSUMER_METADATA_V0.write(buffer, metadata.struct);
```

Comment:
I'm not sure I understand how versioning will work for this data format. Do we need to also include a schema version number as part of this `write()` call? Otherwise, how would we introduce a V1? It doesn't look like this is handled in PartitionAssignmentProtocol since the version of the schema isn't exposed anywhere in this interface.